File: learn-intermediate-python-3/4-OOP/project/school-catalogue.py
class School:
    def __init__(self, name, level, numberOfStudents):
        self.name = name
        self.level = level
        self.numberOfStudents = numberOfStudents

    def get_name(self):
        return self.name
    
    def get_level(self):
        return self.level
    
    def get_numberOfStudents(self):
        return self.numberOfStudents

    def set_numberOfStudents(self, numberOfStudents):
        self.numberOfStudents = numberOfStudents

    def __repr__(self):
        return "A {level} school named {name} with {numberOfStudents} students. ".format(level = self.level, name = self.name, numberOfStudents = self.numberOfStudents)

a = School('School A', 'high', 101)
print(a)
# a.set_numberOfStudents(5000)
# print(a)

class PrimarySchool(School):
    def __init__(self, name, numberOfStudents, pickupPolicy):
        super().__init__(name, 'primary', numberOfStudents)
        self.pickupPolicy = pickupPolicy

    def __repr__(self):
        return super().__repr__() + "The pickup policy is {pickupPolicy}".format(pickupPolicy = self.pickupPolicy)

    def get_pickupPolicy(self):
        return self.pickupPolicy

b = PrimarySchool('School B', 1000, 'NO')
print(b)

class HighSchool(School):
    def __init__(self, name, numberOfStudents, sportsTeams = []):
        super().__init__(name, 'high', numberOfStudents)
        self.sportsTeams = sportsTeams
    
    def get_sportsTeams(self):
        return self.sportsTeams

    def __repr__(self):
        return super().__repr__() + "Sports Teams: {sportsTeamsList}".format(sportsTeamsList = self.sportsTeams)

mvhs = HighSchool("MVHS", 108, ['Football', 'Baseball'])
print(mvhs)

File: backend-engineer/7_javascript-syntax-pt3/1.classes/school-catalogue.js
class School {
    constructor(name, level, numberOfStudents) {
        this._name = name;
        this._level = level;
        this._numberOfStudents = numberOfStudents;
    }

    get name() {
        return this._name;
    }

    get level() {
        return this._level;
    }

    get numberOfStudents() {
        return this._numberOfStudents;
    }

    set numberOfStudents(numberOfStudents) {
        let isNum = typeof numberOfStudents;
        if (isNum != 'number') {
            console.log('Invalid input: numberOfStudents must be set to a Number.');
        } else {
            this._numberOfStudents = numberOfStudents;
        }
    }

    quickFacts() {
        return console.log(`${this._name} educates ${this._numberOfStudents} students at the ${this._level} school level.`);
    };

    static pickSubstituteTeacher(substituteTeachers) {
        const random = Math.floor(Math.random() * (substituteTeachers.length));
        return substituteTeachers[random];
    }

}// END class School


class PrimarySchool extends School {
    constructor(name, numberOfStudents, pickupPolicy) {
        super(name, 'Primary', numberOfStudents);
        this._pickupPolicy = pickupPolicy;
    }

    get pickupPolicy() {
        return this._pickupPolicy;
    }

}// END class PrimarySchool


class MiddleSchool extends School {
    constructor(name, numberOfStudents) {
        super(name, 'Middle', numberOfStudents);
    }
}// END class MiddleSchool


class HighSchool extends School {
    constructor(name, numberOfStudents, sportsTeams) {
        super(name, 'High', numberOfStudents);

        this._sportsTeams = sportsTeams;
    }

    get sportsTeams() {
        return this._sportsTeams;
    }
}// END class HighSchool

const lorraineHansbury = new PrimarySchool('Lorraine Hansbury', 514, 'Students must be picked up by a parent, guardian, or a family member over the age of 13.');
lorraineHansbury.quickFacts();

const mvisd = new MiddleSchool('Medina Valley', 108);
mvisd.quickFacts();

const alSmith = new HighSchool('Al E. Smith', 415, ['Baseball', 'Basketball', 'Volleyball', 'Track and Field']);
alSmith.quickFacts();



// let sub = School.pickSubstituteTeacher(['Jamal Crawford', 'Lou Williams', 'J. R. Smith', 'James Harden', 'Jason Terry', 'Manu Ginobli']);
// console.log(sub);

// let teams = alSmith.sportsTeams;
// console.log(teams);
